### ISSUE TITLE
Possible solution to allowing dynamic include

### ISSUE THREAD

**Issue by smaili** [jadejs/jade#2030]
There's so many issues that are open regarding this I honestly am not sure which one (if any) is the main/active one so I'm creating a solution that I hope may offer a solution :)

As I understand correctly, the reason for currently not having the ability for dynamic includes is due to compilation.  Assuming that is correct, I would like to propose allowing the use of dynamic include with the constraint that you must give some sort of compiler annotation that provides a list of all possible values of that dynamic include.

For example, say I have the template `main.jade`:

```Jade
//- main.jade
-
    var name = '';
    if ( input === 'a' ) name = 'a';
    else if ( input === 'b' ) name = 'b';
    else if ( input === 'c' ) name = 'c';
    else name = 'd';

include /templates/#{ name }
```

In order to let the compiler know all of the possible files that may be included, we just include some kind of special annotation.
```Jade
//- main.jade
-
    var name = '';
    if ( input === 'a' ) name = 'a';
    else if ( input === 'b' ) name = 'b';
    else if ( input === 'c' ) name = 'c';
    else name = 'd';

//- Give the compiler key/value list of possible values for each variable used in the include
include /templates/#{ name }//-{ name: [ 'a', 'b', 'c', 'd' ] }
```

**Comment by codewreaker** [comment]
I agree with this post. I am also suggesting this option for angular
```` html
    section(data-ng-repeat="menu in menus")
        include ./{{menu.url
````

**Comment by smaili** [comment]
Thanks @codewreaker!  Your code inspired me to think of another possible format that is actually more consistent with Jade:

```Jade
//- main.jade
-
    var name = '';
    if ( input === 'a' ) name = 'a';
    else if ( input === 'b' ) name = 'b';
    else if ( input === 'c' ) name = 'c';
    else name = 'd';

//- Give the key/value list as an attribute list
include( name=[ 'a', 'b', 'c', 'd' ] ) /templates/#{ name }
```

**Comment by codewreaker** [comment]
Just a question smaili, what becomes input in this case?

**Comment by acidjazz** [comment]
![image](https://cloud.githubusercontent.com/assets/967369/9105740/733ba604-3bcf-11e5-845f-06bc3b16a542.png)


I can specify 10+ examples where this would make my life so much easier with my projects

**Comment by acidjazz** [comment]
Here is one I am working on right now, data.project is a massive object populating projects on an HTML page, imagine how awesome this code would look like if I had dynamic includes!

```jade

  - var project = data.project.threedr

  .project
    include _header
    .body
      include threedr
    include _credits

  - var project = data.project.versus

  .project
    include _header
    .body
      include versus
    include _credits

  include ../inc/bottom

```

**Comment by ForbesLindesay** [comment]
This idea could actually work moderately ok.  e.g. we could have:

```jade
include ./templates/{name:a,b,c}.jade
```

Which could automatically be expanded into:

```jade
case name
  when 'a'
    include ./templates/a.jade
  when 'b'
    include ./templates/b.jade
  when 'c'
    include ./templates/c.jade
  default
    - throw new Error('"name" must be "a", "b" or "c"')
```

Thoughts:

 1. I'd prefer the possible values of `name` to be specified directly with the input `name`.  That way the tight coupling between the two is more apparent.
 1. Don't reuse the `#{...}` syntax because the content is much more restricted, and not plain JavaScript.  In this environment I think something like `{...}` would probably work fine.
 1. Although it might be possible to do something fairly similar for extends, it would have a much higher overhead, because it would require building the entire template multiple times.

**Comment by ForbesLindesay** [comment]
If anyone wants to have a go at implementing this:

1. Check out [jade-filters/lib/handle-filters.js](https://github.com/jadejs/jade-filters/blob/master/lib/handle-filters.js) for an example of a transformer that is applied to a jade AST.
2. Check out [jadejs/jade-parser](https://github.com/jadejs/jade-parser/blob/8d744cdb46e4ffca6e31450b796a1cf08009ba36/index.js#L614-L631) for examples of how to create an "include" node and a case/when/code node that can be consumed by the later stages of the compilation pipeline.
3. In [jadejs/jade](https://github.com/jadejs/jade/blob/36590e40e4dd1f2c9f59344c05d32387113706f8/lib/index.js#L56) You will need to replace:

```diff
- parse: parse,
+ parse: function (tokens, filename) {
+   var ast = parse(tokens, filename);
+   ast = handleDynamicIncludes(ast);
+   return ast;
+ }
```

**Comment by codewreaker** [comment]
This is so true. I hear they are working on this for the next version.

On Thu, Aug 6, 2015 at 8:10 AM, kevin olson <notifications@github.com>
wrote:

> [image: image]
> <https://cloud.githubusercontent.com/assets/967369/9105740/733ba604-3bcf-11e5-845f-06bc3b16a542.png>
>
> I can specify 10+ examples where this would make my life so much easier
> with my projects
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jadejs/jade/issues/2030#issuecomment-128272950>.
>



-- 
Kind Regards,
Israel Agyeman-Prempeh

**Comment by nullivex** [comment]
I am working on a system where I need to be able to render remote jade files on to the main and I guess jade is handling the path inclusion differently than I expected.

Most importantly I want to dynamically point at my layout.jade with a constant.

```jade
extends MY_CONSTANT_ROOT + '/layout.jade'
```

Would this discussion pertain to that or is it different?

**Comment by nullivex** [comment]
My goal is that I want to prevent my sub views from doing something lame like this.

```jade
extends ../../../../interfaces/admin/view/layout
```

Where it becomes a guessing game at what to point to. Am I barking up the wrong tree?

**Comment by ForbesLindesay** [comment]
@nullivex This has been discussed at length in other issues.  You can use the `basedir` option, but note that the can only be one `basedir` to avoid confusion.  No, there won't be an option to have multiple basedirs as this would be an anti-pattern and lead to terrible code.  No, you can't have fully dynamic include or extends because it's impossible to implement efficiently.  If you really really need this, fork jade and build your own implementation of jade-load and jade-linker with your desired behaviour.

**Comment by nullivex** [comment]
@ForbesLindesay Thanks for the response. I am going to try the `basedir` option and see how it helps me.

Honestly, I was having trouble setting up a google query that would yield any usable response on the subject. Apologies for the double post.

Thanks

**Comment by nullivex** [comment]
@ForbesLindesay Alright, that setting works great. Thanks for pointing that out.

Also I just want to say thank you to Jade. The only template engine I have ever used that didnt make me want to make my own (including a few of my own that still didnt accomplish it).

THANK YOU!!

**Comment by smaudet** [comment]
Sounds awesome!

**Comment by bloadvenro** [comment]
I think you can implement such feature using JADE/PUG API. My quick solution is:

res.render/pug.renderFile('viewfile.pug', {dynamicHtmlPiece: pug.render('include /path/to/file/' + file)}

**Comment by acidjazz** [comment]
@bloadvenro that has been suggested before and is a good work-around for some situations, but unfortunately most of us would love the functionality inside Jade/Pug itself.

**Comment by nullivex** [comment]
@bloadvenro Yes, it is acceptable to work in chunks for single page situations.

However, in my case I was looking to use dynamically included layout pages in my blocks which wasnt going to allow me to render the content prior to pug rendering.

**Comment by Bamieh** [comment]
plus plus this is needed,

i have a critical css file for every page i have, and i want to have one include in the layout to include page specific ones, but i am unable to achieve this. i have to repeat the code in every view.

**Comment by hilleer** [comment]
Is there any news on this issue? 

I have a similar case as many others, where I am iterating an array, and need to include a .md file for each array element. example:

Array:
`var list = ['category1', 'category2', 'category3'];`

.Md files: 
```
path/category1.md
path/category2.md
path/category3.md
```

So I am using mixins, and want to do something like this, when iterating a mixin:
```
mixin category-card(category)

   include:marked "../public/md-files/" + #{category} + ".md"
```

I have checked soooo many of these issues, and yet found no solution. Is it supported yet, or is there any workarounds/plans of releasing such feature? :-)

**Comment by ForbesLindesay** [comment]
There will probably never be a way of doing what you're asking for.  As discussed in this issue, we may in the future be able to offer a very isolated version of this providing you specify up front all possible values of `category`.  I'm going to lock this issue to prevent further "me too" style comments.
